lib/WUI/nhttp/gcode_upload.cpp

```diff
@@ -118,6 +121,27 @@ GcodeUpload::UploadResult GcodeUpload::start(const RequestParser &parser, Upload
         // Missing USB -> Insufficient storage.
         return StatusPage(Status::InsufficientStorage, StatusPage::CloseHandling::ErrorClose, json_errors, "Missing USB drive");
     }
+    struct statvfs svfs;
+    if (statvfs(fname, &svfs) || (svfs.f_bavail * svfs.f_frsize * svfs.f_bsize) < *parser.content_length) {
+        file.reset();
+        remove(fname);
+        return StatusPage(Status::InsufficientStorage, StatusPage::CloseHandling::ErrorClose, json_errors, "USB drive full");
+    }
+    const size_t csize = svfs.f_frsize * svfs.f_bsize;
+    /* Pre-allocate the area needed for the file to prevent frequent metadata updates during the upload.
+       We do the pre-allocation incrementally cluster-by-cluster to minimize the chance of stalling
+       the media and blocking other tasks potentially waiting for the file system lock.
+       The file size doesn't need to be exactly the size of the uploaded content. We will truncate
+       the file later on. */
+    for (unsigned long sz = csize; sz < *parser.content_length; sz += csize) {
+        if (fseek(file.get(), sz, SEEK_SET)) {
```

Comment:
Does `fseek` resize the file incrementally? Or is it some hidden "feature" of our implementation?

Reply:
In general that depends on the FS driver implementation, but I can hardly imagine a FS driver that could do it atomically.